src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
+        # to failed if we thought the pod was running but it's in some other
+        # state. Otherwise, go with our current state.
+        if phase is None:
+            state.status = LabStatus.FAILED
+            state.pod = PodState.MISSING
+        elif state.status == LabStatus.RUNNING:
+            if phase != PodPhase.RUNNING:
+                state.status = LabStatus.FAILED
+        return state
+
+    async def list_lab_users(self, *, only_running: bool = False) -> list[str]:
+        """List all users with labs.
+
+        Parameters
+        ----------
+        only_running
+            If set to `True`, only list users with running labs, not labs in
+            other states.
+
+        Returns
+        -------
+        list of str
+            Users with labs.
+        """
+        if only_running:
+            return [
+                u
+                for u, s in self._labs.items()
+                if s.state and s.state.status == LabStatus.RUNNING
+            ]
+        else:
+            return [u for u, s in self._labs.items() if s.state]
+
+    async def start(self) -> None:
+        """Synchronize with Kubernetes and start a background refresh task.
+
+        Examine Kubernetes for current user lab state, update our internal
+        data structures accordingly, and then start a background refresh task
+        that does this periodically. (Labs may be destroyed by Kubernetes node
+        upgrades, for example.)
+        """
+        if self._scheduler:
+            msg = "User lab state tasks already running, cannot start again"
+            self._logger.warning(msg)
+            return
+        await self._reconcile_lab_state()
+        self._logger.info("Starting periodic reconciliation task")
+        self._scheduler = Scheduler()
+        await self._scheduler.spawn(self._reconcile_loop())
+        self._logger.info("Starting reaper for spawn monitoring tasks")
+        await self._scheduler.spawn(self._reap_spawners())
+
+    async def stop(self) -> None:
+        """Stop the background refresh task."""
+        if not self._scheduler:
+            msg = "User lab state background tasks were already stopped"
+            self._logger.warning(msg)
+            return
+        self._logger.info("Stopping user lab state background tasks")
+        await self._scheduler.close()
+        self._scheduler = None
+        self._logger.info("Stopping spawning monitor tasks")
+        labs = self._labs
+        self._labs = {}
+        for state in labs.values():
+            await state.monitor.cancel()
+
+    async def _delete_lab(
+        self,
+        username: str,
+        state: UserLabState,
+        events: AsyncMultiQueue[Event],
+        *,
+        start_progress: int = 25,
+        end_progress: int = 100,
     ) -> None:
         """Delete the user's lab and namespace.
 
-        Currently, this just deletes the namespace and lets that delete the
```

Comment:
Huh.  Was that actually true up until now?

Reply:
No, the comment was just wrong.